src/geometry/GeoTrackView.i.hh

```diff
@@ -89,34 +89,33 @@ CELER_FUNCTION void GeoTrackView::find_next_step()
 
 //---------------------------------------------------------------------------//
 /*!
- * Move to the next boundary.
+ * Move to the next boundary and update volume accordingly
  */
 CELER_FUNCTION void GeoTrackView::move_next_step()
 {
-    vgstate_ = vgnext_;
-    vgnext_.Clear();
-
     // Move the next step plus an extra fudge distance
     axpy(next_step_, dir_, &pos_);
+    this->move_next_volume();
 }
 
 //---------------------------------------------------------------------------//
 /*!
- * Get the volume ID in the current cell.
+ * Update state to next volume
  */
-CELER_FUNCTION VolumeId GeoTrackView::volume_id() const
+CELER_FUNCTION void GeoTrackView::move_next_volume()
 {
-    REQUIRE(this->boundary() == Boundary::inside);
-    return VolumeId{this->volume().id()};
+    vgstate_ = vgnext_;
+    vgnext_.Clear();
```

Comment:
I thought the `GlobalLocator::IsStepSame` had to be used instead? What happened to that? 😕

Reply:
There is no `IsStepSame()` function, so I believe you're talking about `GlobalLocator::HasSamePath()`.  This function is a bit expensive, so we should avoid it unless it is really necessary. It is probably useful for curved trajectories, since next_step assumes a straight line.  But for `LinearPropagator` we don't need to worry about it, as long as we keep `next_step` and `vgnext` and the particle is moving along the same direction.

Reply:
There is no `IsStepSame()` function, so I believe you're talking about `GlobalLocator::HasSamePath()`. This function is a bit expensive, so we should avoid it unless it is really necessary. It is probably useful for curved trajectories, since next_step assumes a straight line. But for `LinearPropagator` we don't need to worry about it, as long as we keep `next_step` and `vgnext` and the particle is moving along the same direction.